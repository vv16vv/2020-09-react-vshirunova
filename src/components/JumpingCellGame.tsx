import React from "react";
import {Field} from "./Field";

interface GameProp {
    width: number;
    height: number;
    frequency?:number;
}

interface GameState {
    frequency: number;
    x: number;
    y: number;
    timerId?: ReturnType<typeof setInterval>;
}

class NegativeCoord extends Error {
    constructor(coord: number) {
        super(`Coordinate ${coord} cannot be 0 or negative`);
    }
}

class TooBigCoord extends Error {
    constructor(coord: number, max: number) {
        super(`Coordinate ${coord} bigger than a maximum ${max}`);
    }
}

export class JumpingCellGame extends React.Component<GameProp, GameState> {
    constructor(props: GameProp) {
        super(props);
        console.log(`JumpingCellGame - constructor: props = ${JSON.stringify(props)}`);
        this.state = {
            frequency: this.props.frequency || 2000,
            x: 1,
            y: 1,
        };
    }

    handleClick = (x: number, y: number) => {
        console.log(`JumpingCellGame - handleClick: x = ${x}, y = ${y}`);
        if (this.state.x === x && this.state.y === y) {
            this.increaseFrequency()
        }
    }

    handleJump = () => {
        console.log(`JumpingCellGame - handleJump`);
        const {width, height} = this.props;
        const newX = this.nextRandomCoord(width);
        const newY = this.nextRandomCoord(height);
        this.setState({
            x: newX,
            y: newY
        })
    }

    nextRandomCoord = (quantity: number): number => {
        return Math.floor(Math.random() * quantity + 1)
    }

    assertCoordIsOk = (coord: number, max: number): number => {
        if (coord <= 0) throw new NegativeCoord(coord);
        if (coord > max) throw new TooBigCoord(coord, max);
        return coord
    }

    increaseFrequency = () => {
        console.log(`JumpingCellGame - increaseFrequency`);
        const {frequency} = this.state;
        this.setState({
                frequency: Math.max(10, Math.round(frequency / 2))
            }
        )
    }

    componentDidMount() {
        console.log(`JumpingCellGame - componentDidMount`);
        if (this.state.timerId === undefined) {
            const timerId = setInterval(this.handleJump, this.state.frequency);
            this.setState({
                timerId: timerId
            })
        }
    }

    componentWillUnmount() {
        console.log(`JumpingCellGame - componentWillUnmount`);
        if (this.state.timerId !== undefined) {
            clearInterval(this.state.timerId);
            this.setState({
                timerId: undefined
            })
        }
    }

    componentDidUpdate(prevProps: Readonly<GameProp>, prevState: Readonly<GameState>) {
        console.log(`JumpingCellGame - componentDidUpdate: prevProps = ${JSON.stringify(prevProps)}, prevState = ${JSON.stringify(prevState)}`);
        const {frequency: oldFreq} = prevState;
        const {frequency: currFreq, timerId: currTimeId} = this.state;
        if (oldFreq !== currFreq && currTimeId !== undefined) {
            clearInterval(currTimeId)
            const timerId = setInterval(this.handleJump, currFreq);
            this.setState({
                timerId: timerId
            })
        }

    }

    getSnapshotBeforeUpdate(prevProps: Readonly<GameProp>, prevState: Readonly<GameState>): any | null {
        console.log(`JumpingCellGame - getSnapshotBeforeUpdate: prevProps = ${JSON.stringify(prevProps)}`)
        console.log(`JumpingCellGame - getSnapshotBeforeUpdate: ${JSON.stringify(prevState)}`)
        return null;
    }

    render() {
        const x = this.assertCoordIsOk(this.state.x, this.props.width)
        const y = this.assertCoordIsOk(this.state.y, this.props.height)
        console.log(`JumpingCellGame - render`);
        return <Field
            width={this.props.width}
            height={this.props.height}
            filledCells={[{x: x, y: y}]}
            clickHandler={this.handleClick}
        />;
    }
}
