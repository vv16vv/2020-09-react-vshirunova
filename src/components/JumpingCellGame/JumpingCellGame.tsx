import React from "react";
import {Field} from "cmp/Field";
import {SettingsFormResult} from "cmp/SettingsForm";
import {StyledBlock, StyledButton} from "styled/StyledComponents";
import {CenteredLabel} from "styled/StyledTextComponents";

interface GameState {
    frequency: number;
    width: number;
    height: number;
    x: number;
    y: number;
    jumps: number;
    clicks: number;
    timerId?: ReturnType<typeof setInterval>;
}

class NegativeCoord extends Error {
    constructor(coord: number) {
        super(`Coordinate ${coord} cannot be 0 or negative`);
    }
}

class TooBigCoord extends Error {
    constructor(coord: number, max: number) {
        super(`Coordinate ${coord} bigger than a maximum ${max}`);
    }
}

const initialState: GameState = {
    x: 1,
    y: 1,
    width: 10,
    height: 10,
    jumps: 0,
    clicks: 0,
    frequency: 1000
} as const;

export class JumpingCellGame extends React.Component<SettingsFormResult, GameState> {
    constructor(props: SettingsFormResult) {
        super(props);
        console.log(`JumpingCellGame - constructor: props = ${JSON.stringify(props)}`);
        this.state = {
            ...initialState,
            ...props,
        };
    }

    handleClick = (x: number, y: number) => {
        console.log(`JumpingCellGame - handleClick: x = ${x}, y = ${y}`);
        this.incrementClicks()
        if (this.state.x === x && this.state.y === y) {
            this.increaseFrequency()
            this.incrementJumps()
            this.handleJump()
        }
    }

    handleJump = () => {
        console.log(`JumpingCellGame - handleJump`);
        const {width, height} = this.state;
        const newX = this.nextRandomCoord(width);
        const newY = this.nextRandomCoord(height);
        this.setState({
            x: newX,
            y: newY
        })
    }

    nextRandomCoord = (quantity: number): number => {
        return Math.floor(Math.random() * quantity + 1)
    }

    assertCoordIsOk = (coord: number, max: number): number => {
        if (coord <= 0) throw new NegativeCoord(coord);
        if (coord > max) throw new TooBigCoord(coord, max);
        return coord
    }

    increaseFrequency = () => {
        console.log(`JumpingCellGame - increaseFrequency`);
        const {frequency} = this.state;
        this.setState({
                frequency: Math.max(10, Math.round(frequency / 2))
            }
        )
    }

    incrementClicks = () => {
        console.log(`JumpingCellGame - incrementClicks`);
        const {clicks} = this.state;
        this.setState({
                clicks: clicks + 1
            }
        )
    }

    incrementJumps = () => {
        console.log(`JumpingCellGame - incrementJumps`);
        const {jumps} = this.state;
        this.setState({
                jumps: jumps + 1
            }
        )
    }

    resetHandle = () => {
        console.log(`JumpingCellGame - resetHandle`);
        this.setState({
                ...initialState,
                frequency: this.props.frequency,
                width: this.props.width,
                height: this.props.height,
            }
        )
    }

    componentDidMount() {
        console.log(`JumpingCellGame - componentDidMount`);
        if (this.state.timerId === undefined) {
            const timerId = setInterval(this.handleJump, this.state.frequency);
            this.setState({
                timerId: timerId
            })
        }
    }

    componentWillUnmount() {
        console.log(`JumpingCellGame - componentWillUnmount`);
        if (this.state.timerId !== undefined) {
            clearInterval(this.state.timerId);
            this.setState({
                timerId: undefined
            })
        }
    }

    checkFieldWidth = (newWidth: number, oldWidth: number) => {
        const {x} = this.state;
        if (newWidth < oldWidth && x > newWidth) {
            this.setState({
                x: newWidth,
                width: newWidth,
            })
        }
    }

    checkFieldHeight = (newHeight: number, oldHeight: number) => {
        const {y} = this.state;
        if (newHeight < oldHeight && y > newHeight) {
            this.setState({
                y: newHeight,
                height: newHeight,
            })
        }
    }

    checkFrequencyReset = (newFreq: number, oldFreq: number) => {
        if (newFreq !== oldFreq) {
            const {timerId: currTimeId} = this.state;
            clearInterval(currTimeId as NodeJS.Timeout)
            const timerId = setInterval(this.handleJump, newFreq);
            this.setState({
                frequency: newFreq,
                timerId: timerId
            })
        }
    }

    checkFrequencyChange = (newFreq: number, oldFreq: number) => {
        const {timerId: currTimeId} = this.state;
        if (newFreq !== oldFreq && currTimeId !== undefined) {
            clearInterval(currTimeId as NodeJS.Timeout)
            const timerId = setInterval(this.handleJump, newFreq);
            this.setState({
                timerId: timerId
            })
        }
    }

    componentDidUpdate(prevProps: Readonly<SettingsFormResult>, prevState: Readonly<GameState>) {
        console.log(`JumpingCellGame - componentDidUpdate: prevProps = ${JSON.stringify(prevProps)}, prevState = ${JSON.stringify(prevState)}`);
        this.checkFieldWidth(this.props.width, prevState.width)
        this.checkFieldHeight(this.props.height, prevState.height)
        this.checkFrequencyReset(this.props.frequency, prevProps.frequency)
        this.checkFrequencyChange(this.state.frequency, prevState.frequency)
    }

    getSnapshotBeforeUpdate(prevProps: Readonly<SettingsFormResult>, prevState: Readonly<GameState>): any | null {
        console.log(`JumpingCellGame - getSnapshotBeforeUpdate: prevProps = ${JSON.stringify(prevProps)}`)
        console.log(`JumpingCellGame - getSnapshotBeforeUpdate: ${JSON.stringify(prevState)}`)
        return null;
    }

    render() {
        const {width, height} = this.state;
        const x = this.assertCoordIsOk(this.state.x, width)
        const y = this.assertCoordIsOk(this.state.y, height)
        console.log(`JumpingCellGame - render`);
        return <>
            <StyledBlock>
                <Field
                    width={width}
                    height={height}
                    filledCells={[{x: x, y: y}]}
                    clickHandler={this.handleClick}
                />
            </StyledBlock>
            <StyledBlock>
                <CenteredLabel>{`Jumps: ${this.state.jumps}`}</CenteredLabel>
                <CenteredLabel>{`Clicks: ${this.state.clicks}`}</CenteredLabel>
                <StyledButton onClick={this.resetHandle}>Reset</StyledButton>
            </StyledBlock>
        </>;
    }
}
