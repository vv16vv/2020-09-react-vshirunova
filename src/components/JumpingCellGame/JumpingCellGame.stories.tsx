import React from "react";
import {number, withKnobs} from "@storybook/addon-knobs";
import {JumpingCellGame} from "./JumpingCellGame";
import {ErrorBoundary} from "cmp/ErrorBoundary";

export default {
    title: "Jumping Cell Game Story",
    decorators: [withKnobs],
}

export const JumpingCellGameStory: React.FC<{}> = () => {
    const options = {
        range: true,
        min: 5,
        max: 20,
        step: 1
    };
    const optionsFreq = {
        range: true,
        min: 2000,
        max: 10000,
        step: 500
    };
    const width = number("Width", 10, options);
    const height = number("Height", 10, options);
    const frequency = number("Frequency", 5000, optionsFreq);
    return <ErrorBoundary>
        <JumpingCellGame width={width} height={height} frequency={frequency}/>
    </ErrorBoundary>
};