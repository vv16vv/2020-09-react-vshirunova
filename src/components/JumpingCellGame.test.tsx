import renderer from "react-test-renderer";
import React from "react";
import {JumpingCellGame} from "./JumpingCellGame";

describe("Jumping cell game", () => {

    const width = 3;
    const height = 4;

    it("should be rendered", () => {
        expect(renderer.create(
            <JumpingCellGame
                width={width}
                height={height}
            />
            ).toJSON()
        ).toMatchSnapshot();
    });

});