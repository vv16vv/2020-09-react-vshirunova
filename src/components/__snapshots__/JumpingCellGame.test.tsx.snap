// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Jumping cell game should be rendered 1`] = `
<table>
  <tbody>
    <tr>
      <td
        id="id1-1"
        onClick={[Function]}
      >
        [X] 1, 1
      </td>
      <td
        id="id2-1"
        onClick={[Function]}
      >
        [ ] 2, 1
      </td>
      <td
        id="id3-1"
        onClick={[Function]}
      >
        [ ] 3, 1
      </td>
    </tr>
    <tr>
      <td
        id="id1-2"
        onClick={[Function]}
      >
        [ ] 1, 2
      </td>
      <td
        id="id2-2"
        onClick={[Function]}
      >
        [ ] 2, 2
      </td>
      <td
        id="id3-2"
        onClick={[Function]}
      >
        [ ] 3, 2
      </td>
    </tr>
    <tr>
      <td
        id="id1-3"
        onClick={[Function]}
      >
        [ ] 1, 3
      </td>
      <td
        id="id2-3"
        onClick={[Function]}
      >
        [ ] 2, 3
      </td>
      <td
        id="id3-3"
        onClick={[Function]}
      >
        [ ] 3, 3
      </td>
    </tr>
    <tr>
      <td
        id="id1-4"
        onClick={[Function]}
      >
        [ ] 1, 4
      </td>
      <td
        id="id2-4"
        onClick={[Function]}
      >
        [ ] 2, 4
      </td>
      <td
        id="id3-4"
        onClick={[Function]}
      >
        [ ] 3, 4
      </td>
    </tr>
  </tbody>
</table>
`;
